### PR TITLE
Generalize units in slice plugin

### PR DESCRIPTION
**Pull request by kecnry**
### Description

This pull request generalizes the slice plugin for:
* future support for display units in cubeviz (needing to alternate between wavelength and frequency)
* [downstream use in lcviz](https://github.com/spacetelescope/lcviz/pull/85)

It does so by:
* replacing (with deprecation) any user-exposed mention of "wavelength" with either "value" (in the API) or a dynamic string ("wavelength", "frequency", "time", etc)
* allowing for dynamic units in the plugin UI
* ~allowing overriding the plugin header text when inherited by downstream apps~ (moved to #2709)
* refactoring so setting the slice can be overridden downstream (in cases where it isn't the last index as-is in cubeviz)
* removing "(spectral axis)" from the tooltip description for the slice viewer tool
* slice indicator tool sets value via message (generalized with value instead of wavelength) instead of assuming the helper has `select_wavelength`.
* moving `slice_indicator` property to a mixin so that it can be used by other viewers that are not spectral.

Coming as a follow-up PR:
* breaking the assumption of a single cube with a shape the same as the length of the 1d data (spectrum).

**Before PR**:

Screen Shot 2024-02-14 at 1 45 01 PM

**After PR**:

Screen Shot 2024-02-14 at 1 44 27 PM

### Change log entry

- [x] Is a change log needed? If yes, is it added to `CHANGES.rst`? If you want to avoid merge conflicts,
  list the proposed change log here for review and add to `CHANGES.rst` before merge. If no, maintainer
  should add a `no-changelog-entry-needed` label.

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [x] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [ ] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Did the CI pass? If not, are the failures related?
- [x] Is a milestone set? Set this to bugfix milestone if this is a bug fix and needs to be released ASAP; otherwise, set this to the next major release milestone.
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)?

### CODE REVIEW

File: jdaviz/configs/cubeviz/plugins/slice/slice.vue [hunk 1]
```diff
@@ -1,6 +1,6 @@
 <template>
   <j-tray-plugin
-    description='Select slice (or wavelength) of the cube to show in the image viewers and highlighted in the spectrum viewer.  The slice can also be changed interactively in the spectrum viewer by activating the slice tool.'
+    :description="docs_description || 'Select slice of the cube to show in the image viewers.  The slice can also be changed interactively in the spectrum viewer by activating the slice tool.'"
```
Comment (pllim) [inline jdaviz/configs/cubeviz/plugins/slice/slice.vue:3]: This doesn't seem related?

File: jdaviz/configs/cubeviz/plugins/slice/slice.vue [hunk 2]
```diff
@@ -14,16 +14,16 @@
...
                 persistent-hint>
               </v-switch>
             </v-row>
             <v-row>
               <v-switch
-                label="Show Wavelength"
-                hint="Show slice wavelength in label to right of indicator."
-                v-model="show_wavelength"
+                label="Show Value"
+                :hint="'Show slice '+value_label.toLowerCase()+' in label to right of indicator.'"
```
Comment (pllim) [inline jdaviz/configs/cubeviz/plugins/slice/slice.vue:25]: I don't understand the need for `toLowerCase` here.

File: jdaviz/configs/cubeviz/plugins/slice/slice.vue (hunk 1 above)
Comment (kecnry) [reply jdaviz/configs/cubeviz/plugins/slice/slice.vue:3]: It is, it's covered in the bullet "allowing overriding the plugin header text when inherited by downstream apps" and coupled with setting `docs_description` as an empty traitlet in the plugin template mixin.

File: jdaviz/configs/cubeviz/plugins/slice/slice.vue (hunk 2 above)
Comment (kecnry) [reply jdaviz/configs/cubeviz/plugins/slice/slice.vue:25]: Because we want lower case here (not "Show slice Wavelength in label"), but we want title case in the actual input widget, without defining two different strings on the python-side.

**Review by pllim — COMMENTED** [review]
In principle, looks fine. Just a couple of minor comments above.

File: jdaviz/configs/cubeviz/plugins/slice/slice.vue (hunk 2 above)
Comment (pllim) [reply jdaviz/configs/cubeviz/plugins/slice/slice.vue:25]: Okay, just to make sure, this does not affect string representation of the unit, right? Because "MJy" is not the same as "mjy".

Comment (kecnry) [reply jdaviz/configs/cubeviz/plugins/slice/slice.vue:25]: now, this is the label (Wavelength, Frequency, Time, etc), not the unit.

File: jdaviz/configs/cubeviz/plugins/slice/slice.vue (hunk 1 above)
Comment (kecnry) [reply jdaviz/configs/cubeviz/plugins/slice/slice.vue:3]: would you rather I split this into a separate PR (and then add the hookups for all the plugins there)?

**Review by pllim — APPROVED** [review]
LGTM. Thanks for the clarifications!

File: jdaviz/configs/cubeviz/plugins/slice/slice.vue (hunk 1 above)
Comment (kecnry) [reply jdaviz/configs/cubeviz/plugins/slice/slice.vue]: moved (and applied to all plugins) in https://github.com/spacetelescope/jdaviz/pull/2709

**Review by gibsongreen — APPROVED** [review]
Very useful for myself to see both for immediate implementation and long term planning. I am excited to see the follow-up PR(s)!
